development/cpp/internal_rendering_architecture.rst

```diff
@@ -0,0 +1,600 @@
+
+**Other GLSL shaders for the Compatibility backend:**
+
+- `drivers/gles3/shaders/ <https://github.com/godotengine/godot/blob/master/drivers/gles3/shaders/>`__
+
+2D and 3D rendering separation
+------------------------------
+
+.. note::
+
+    Only available in the Clustered Forward and Forward Mobile backends, not the
+    Compatibility backend. Multiple Viewports can still be used to emulate this
+    when using the Compatibility backend, or to perform 2D resolution scaling.
+
+2D and 3D are rendered to separate buffers to allow for 3D resolution scaling
+without requiring the use of multiple Viewport nodes (which has greater
+performance and memory usage overhead).
+
+3D resolution scaling is performed differently depending on whether bilinear or
+FSR 1.0 scaling is used. When bilinear scaling is used, no special upscaling
+shader is run. Instead, the viewport's texture is stretched and displayed with a
+linear sampler (which makes the filtering happen directly on the hardware). This
+allows maximizing the performance of bilinear 3D scaling.
+
+The ``configure()`` function in RenderSceneBuffersRD reallocates the 2D/3D
+buffers when the resolution or scaling changes.
+
+Dynamic resolution scaling isn't supported yet, but is planned in a future Godot
+release.
+
+**2D and 3D rendering buffer configuration C++ code:**
+
+- `servers/rendering/renderer_rd/storage_rd/render_scene_buffers_rd.cpp <https://github.com/godotengine/godot/blob/master/servers/rendering/renderer_rd/storage_rd/render_scene_buffers_rd.cpp>`__
+
+2D rendering techniques
+-----------------------
+
+2D light rendering is performed in a single pass to allow for better performance
+with large amounts of lights.
+
+The clustered and mobile backends don't feature 2D batching yet, but it's
+planned for a future release.
+
+The compatibility backend features 2D batching to improve performance, which is
+especially noticeable with lots of text on screen.
+
+MSAA can be enabled in 2D to provide line and polygon antialiasing, but FXAA
+will never affect 2D rendering as it would look too blurry.
+
+A 2D signed distance field representing the viewport is automatically generated
+if an user shader requests it. This can be used for various effects in custom
+shaders, such as 2D global illumination.
+
+**2D SDF generation GLSL shader:**
+
+- `https://github.com/godotengine/godot/blob/master/servers/rendering/renderer_rd/shaders/canvas_sdf.glsl <servers/rendering/renderer_rd/shaders/canvas_sdf.glsl>`__
+
+3D rendering techniques
+-----------------------
+
+Batching and instancing
+^^^^^^^^^^^^^^^^^^^^^^^
+
+In the clustered and mobile backends, Vulkan instancing is used to group
+rendering of identical objects for performance. This is not as fast as static
+mesh merging, but it still allows instances to be culled individually.
+
+In the compatibility backend, a more limited form of 3D instancing is also used.
+**TODO:** Check if this is the case?
```

Comment:
@clayjohn Does the OpenGL backend in `master` use some form of automatic instancing in 3D? I haven't followed recent developments, so I don't remember if this is different from what's done in 3.x.